File: cpp/081-090/Remove Duplicates from Sorted List II.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cmath>
#include <stack>
#include <string>

using namespace std;


struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
 
class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        
        ListNode *result = new ListNode(INT_MIN);
        ListNode *_head = result;
        int prev(INT_MIN);
        
        while(head)
        {
            if(head->val != prev && ( (!head->next) || head->val != head->next->val) )
            {
                _head->next = head;
                _head = head;
            }
            
            prev = head->val;
            head = head->next;
        }
        _head->next = NULL;
        return result->next;
        
    }
    
    
};

int main()
{
    Solution s;
    ListNode *n1 = new ListNode(1);
    ListNode *n2 = new ListNode(2);
    ListNode *n3 = new ListNode(2);
    n1->next = n2;
    n2->next = n3;
    
    ListNode *head = s.deleteDuplicates(n1);
    
    while(head)
    {
        cout << head->val << " ";
        head = head->next;
    }
}

File: python/081-090/Remove Duplicates from Sorted List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def deleteDuplicates(self, head):
        nhead = ListNode(0)
        prev = -2**31
        res = nhead
        
        while head:
            if head.val != prev and (not head.next or head.val != head.next.val) :
                nhead.next = head
                nhead = head
            prev = head.val
            head = head.next
        
        nhead.next = None
        return res.next